src/find_db.cpp

```diff
@@ -45,12 +48,156 @@ boost::optional<std::string>& testing_find_db_path_override()
+        }
+        else
+        {
+            MIOPEN_LOG_I2("inexact find database search");
+            if(fs::exists(root_path) && fs::is_directory(root_path))
+            {
+                MIOPEN_LOG_I2("Iterating over find db directory " << root_path.string());
+                std::vector<fs::path> all_files;
+                for(const auto& kinder : fs::recursive_directory_iterator(root_path))
+                {
+                    const auto& filepath = kinder.path();
+                    const auto& fname    = filepath.string();
+                    if(fs::is_regular_file(kinder) && EndsWith(fname, ".fdb.txt"))
+                        all_files.push_back(filepath);
+                }
+
+                const auto db_id        = handle.GetTargetProperties().DbId();
+                const int real_cu_count = handle.GetMaxComputeUnits();
+                int closest_cu          = std::numeric_limits<int>::max();
+                fs::path best_path;
+                for(const auto& entry : all_files)
+                {
+                    const auto fname = entry.stem().string();
+                    MIOPEN_LOG_I("Checking find db file: " << fname);
+                    // Check for alternate back end same ASIC
+                    if(fname.rfind(base_name, 0) == 0)
+                    {
+                        return entry.string();
+                    }
+                    if(db_id.empty() || !miopen::StartsWith(db_id, "gfx") || real_cu_count == 0)
+                        return std::string();
+                    // Check for alternate ASIC any back end
+                    if(fname.rfind(db_id, 0) == 0)
+                    {
+                        const auto pos = fname.find('_');
+                        int cur_count  = -1;
+                        if(pos != std::string::npos)
+                            cur_count = std::stoi(fname.substr(pos + 1));
+                        else
+                            cur_count = std::stoi(fname.substr(db_id.length()), nullptr, 16);
+                        if(abs(cur_count - real_cu_count) < (closest_cu))
+                        {
+                            best_path  = entry;
+                            closest_cu = abs(cur_count - real_cu_count);
+                        }
+                    }
+                }
+                return best_path.string();
+            }
+            else
+            {
+                MIOPEN_LOG_I("Database directory does not exist");
+                return std::string();
+            }
+        }
+    }();
+    return installed_path;
+}
+#endif
 template <class TDb>
 std::string FindDbRecord_t<TDb>::GetInstalledPath(Handle& handle)
 {
 #if !MIOPEN_DISABLE_SYSDB
-    return GetSystemDbPath() + "/" + handle.GetDbBasename() + "." + GetSystemFindDbSuffix() +
-           ".fdb.txt";
+#if MIOPEN_EMBED_DB
+    return GetInstalledPathEmbed(handle);
+#else
+    return GetInstalledPathFile(handle);
+#endif
```

Comment:
`GetInstalledPath()` is invoked whenever FindDbRecord instance is created. The previous version of `GetInstalledPath()` is very fast and takes ~constant time. The new version adds `boost::filesystem::exists()` at best or loop with file lookup at worst. This would lead to performance drop.

The simplest workaround is caching return value in a static variable.

Reply:
@atamazov Both `GetInstalledPathEmbed` and `GetInstalledPathFile` functions return a static const variable and therefore the above logic should execute only once. Adding another static variable here would be redundant.

Reply:
Yep, I missed those lambdas ;/